File: frontend/dashboard/src/services/integrationService.js
// frontend/dashboard/src/services/integrationService.js
// Path: frontend/dashboard/src/services/integrationService.js
// Usage: Enhanced integration service with comprehensive API methods and error handling

import api from './api';
import { toast } from 'react-hot-toast';

/**
 * Enhanced Integration Service
 * Handles all integration-related API operations with improved error handling and validation
 */
const integrationService = {
  /**
   * Get all integrations for the current client
   * @returns {Promise<Array>} List of integrations
   */
  async getIntegrations() {
    try {
      console.log('Fetching integrations...');
      const response = await api.get('/api/integrations');
      console.log('Integrations response:', response.data);
      return response.data || [];
    } catch (error) {
      console.error('Error fetching integrations:', error);
      
      // Enhanced error handling
      if (error.response) {
        console.error('Server responded with error:', error.response.status, error.response.data);
        
        // Handle specific error cases
        if (error.response.status === 404) {
          console.log('Integrations endpoint not found, returning empty array');
          return [];
        }
      } else if (error.request) {
        console.error('No response received:', error.request);
      } else {
        console.error('Error setting up request:', error.message);
      }
      
      // Return empty array to prevent UI crashes
      return [];
    }
  },

  /**
   * Get available integration providers with setup information
   * @returns {Promise<Array>} List of available providers
   */
  async getAvailableProviders() {
    try {
      const response = await api.get('/api/integrations/available');
      return response.data || [];
    } catch (error) {
      console.error('Error fetching available providers:', error);
      
      // Return default providers if API fails
      return [
        {
          id: 'shopify',
          name: 'Shopify',
          description: 'E-commerce platform integration for products, orders, and customers',
          resource_types: ['products', 'orders', 'customers'],
          fields: [
            {
              name: 'store_domain',
              label: 'Store Domain',
              type: 'text',
              placeholder: 'your-store.myshopify.com',
              required: true,
              help: 'Your Shopify store domain'
            },
            {
              name: 'access_token',
              label: 'Admin API Access Token',
              type: 'password',
              placeholder: 'shpat_...',
              required: true,
              help: 'Admin API Access Token from your Shopify Custom App'
            }
          ]
        }
      ];
    }
  },

  /**
   * Test connection to an integration provider (NEW - Enhanced)
   * @param {Object} connectionData - Connection test data
   * @returns {Promise<Object>} Test result with detailed feedback
   */
  async testConnection(connectionData) {
    try {
      const response = await api.post('/api/integrations/test', connectionData);
      return response.data;
    } catch (error) {
      console.error('Error testing connection:', error);
      
      const errorMessage = error.response?.data?.message || error.message || 'Connection test failed';
      
      return {
        success: false,
        message: errorMessage,
        status_code: error.response?.status || 500
      };
    }
  },

  /**
   * Test connection for an existing integration (Enhanced)
   * @param {string} integrationId - Integration ID
   * @returns {Promise<Object>} Test result
   */
  async testExistingConnection(integrationId) {
    try {
      const response = await api.post(`/api/integrations/${integrationId}/test`);
      return {
        ...response.data,
        integration_id: integrationId
      };
    } catch (error) {
      console.error('Error testing existing connection:', error);
      
      return {
        success: false,
        message: error.response?.data?.message || 'Connection test failed',
        integration_id: integrationId
      };
    }
  },

  /**
   * Create a new integration with enhanced validation
   * @param {Object} integrationData - Integration data
   * @returns {Promise<Object>} Created integration or error details
   */
  async createIntegration(integrationData) {
    try {
      const response = await api.post('/api/integrations', integrationData);
      
      if (response.data.success) {
        return response.data;
      } else {
        throw new Error(response.data.message || 'Failed to create integration');
      }
    } catch (error) {
      console.error('Error creating integration:', error);
      
      const errorMessage = error.response?.data?.message || error.message || 'Failed to create integration';
      
      return {
        success: false,
        message: errorMessage,
        error_code: error.response?.data?.error_code || 'CREATION_ERROR'
      };
    }
  },

  /**
   * Update an existing integration
   * @param {string} integrationId - Integration ID
   * @param {Object} updateData - Data to update
   * @returns {Promise<Object>} Updated integration
   */
  async updateIntegration(integrationId, updateData) {
    try {
      const response = await api.put(`/api/integrations/${integrationId}`, updateData);
      return response.data;
    } catch (error) {
      console.error('Error updating integration:', error);
      
      const errorMessage = error.response?.data?.message || error.message || 'Failed to update integration';
      
      return {
        success: false,
        message: errorMessage
      };
    }
  },

  /**
   * Delete an integration
   * @param {string} integrationId - Integration ID
   * @returns {Promise<Object>} Deletion result
   */
  async deleteIntegration(integrationId) {
    try {
      const response = await api.delete(`/api/integrations/${integrationId}`);
      return response.data;
    } catch (error) {
      console.error('Error deleting integration:', error);
      
      const errorMessage = error.response?.data?.message || error.message || 'Failed to delete integration';
      
      return {
        success: false,
        message: errorMessage
      };
    }
  },

  /**
   * Get details of a specific integration
   * @param {string} integrationId - Integration ID
   * @returns {Promise<Object>} Integration details
   */
  async getIntegration(integrationId) {
    try {
      const response = await api.get(`/api/integrations/${integrationId}`);
      return response.data;
    } catch (error) {
      console.error('Error fetching integration details:', error);
      throw error;
    }
  },

  /**
   * Sync data from an integration provider (Enhanced)
   * @param {string} integrationId - Integration ID
   * @param {Object} options - Sync options (optional)
   * @returns {Promise<Object>} Sync result
   */
  async syncIntegration(integrationId, options = {}) {
    try {
      const response = await api.post(`/api/integrations/${integrationId}/sync`, options);
      return response.data;
    } catch (error) {
      console.error('Error syncing integration:', error);
      
      const errorMessage = error.response?.data?.message || error.message || 'Failed to sync integration';
      
      return {
        success: false,
        message: errorMessage
      };
    }
  },

  /**
   * Get data from an integration provider (Enhanced)
   * @param {string} integrationId - Integration ID
   * @param {string} resourceType - Resource type (products, orders, customers, etc.)
   * @param {Object} params - Query parameters (query, limit, etc.)
   * @returns {Promise<Object>} Integration data with metadata
   */
  async getIntegrationData(integrationId, resourceType, params = {}) {
    try {
      const response = await api.get(`/api/integrations/${integrationId}/data/${resourceType}`, {
        params
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching integration data:', error);
      
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to fetch integration data',
        data: [],
        count: 0
      };
    }
  },

  /**
   * Get data by provider and resource type (Legacy method - Enhanced)
   * @param {string} provider - Provider type
   * @param {string} resourceType - Resource type
   * @param {string} query - Search query
   * @returns {Promise<Array>} Integration data
   */
  async getIntegrationDataByProvider(provider, resourceType, query = '') {
    try {
      // First get integrations to find the right integration ID
      const integrations = await this.getIntegrations();
      const integration = integrations.find(i => i.provider === provider && i.is_active);
      
      if (!integration) {
        console.warn(`No active integration found for provider: ${provider}`);
        return [];
      }
      
      const result = await this.getIntegrationData(integration.integration_id, resourceType, { query });
      return result.data || [];
    } catch (error) {
      console.error('Error fetching integration data by provider:', error);
      return [];
    }
  },

  /**
   * Validate integration credentials before creation (NEW)
   * @param {Object} credentialsData - Credentials to validate
   * @returns {Promise<Object>} Validation result
   */
  async validateCredentials(credentialsData) {
    try {
      // Use the test connection endpoint for validation
      return await this.testConnection(credentialsData);
    } catch (error) {
      return {
        success: false,
        message: 'Credential validation failed',
        error: error.message
      };
    }
  },

  /**
   * Get integration status and health (NEW)
   * @param {string} integrationId - Integration ID
   * @returns {Promise<Object>} Integration status
   */
  async getIntegrationStatus(integrationId) {
    try {
      const integration = await this.getIntegration(integrationId);
      
      return {
        status: integration.status,
        is_active: integration.is_active,
        last_sync: integration.last_sync,
        status_message: integration.status_message,
        health: integration.status === 'connected' ? 'healthy' : 'unhealthy'
      };
    } catch (error) {
      console.error('Error getting integration status:', error);
      return {
        status: 'unknown',
        is_active: false,
        health: 'unknown',
        error: error.message
      };
    }
  },

  /**
   * Bulk operations for multiple integrations (NEW)
   */
  async bulkSync(integrationIds) {
    const results = [];
    
    for (const integrationId of integrationIds) {
      try {
        const result = await this.syncIntegration(integrationId);
        results.push({ integrationId, success: true, result });
      } catch (error) {
        results.push({ 
          integrationId, 
          success: false, 
          error: error.message 
        });
      }
    }
    
    return results;
  },

  /**
   * Get integration metrics and analytics (NEW)
   * @param {string} integrationId - Integration ID
   * @returns {Promise<Object>} Integration metrics
   */
  async getIntegrationMetrics(integrationId) {
    try {
      // This would be implemented when analytics endpoint is available
      const response = await api.get(`/api/integrations/${integrationId}/metrics`);
      return response.data;
    } catch (error) {
      console.error('Error fetching integration metrics:', error);
      
      // Return default metrics structure
      return {
        total_requests: 0,
        successful_requests: 0,
        failed_requests: 0,
        last_sync_time: null,
        data_freshness: 'unknown'
      };
    }
  },

  /**
   * Enhanced error handling with user-friendly messages
   */
  formatError(error, context = '') {
    const baseMessage = context ? `${context}: ` : '';
    
    if (error.response) {
      const status = error.response.status;
      const message = error.response.data?.message || error.message;
      
      switch (status) {
        case 400:
          return `${baseMessage}Invalid request. ${message}`;
        case 401:
          return `${baseMessage}Authentication failed. Please check your credentials.`;
        case 403:
          return `${baseMessage}Access denied. Please check your permissions.`;
        case 404:
          return `${baseMessage}Integration not found.`;
        case 429:
          return `${baseMessage}Rate limit exceeded. Please try again later.`;
        case 500:
          return `${baseMessage}Server error. Please try again later.`;
        default:
          return `${baseMessage}${message || 'An unexpected error occurred.'}`;
      }
    }
    
    return `${baseMessage}${error.message || 'Network error occurred.'}`;
  },

  /**
   * Helper method to show toast notifications (NEW)
   */
  showToast(type, message) {
    if (typeof toast !== 'undefined') {
      switch (type) {
        case 'success':
          toast.success(message);
          break;
        case 'error':
          toast.error(message);
          break;
        case 'loading':
          return toast.loading(message);
        default:
          toast(message);
      }
    } else {
      console.log(`${type.toUpperCase()}: ${message}`);
    }
  }
};

export default integrationService;

File: backend/app/services/integration/integration_service.py
# backend/app/services/integration/integration_service.py
from typing import List, Dict, Any, Optional
import logging
from datetime import datetime
import uuid
from sqlalchemy.orm import Session

from app.domain.integration.entities import Integration, IntegrationSync
from app.repositories.integration_repository import IntegrationRepository, IntegrationSyncRepository
from app.services.integration.providers.provider_factory import ProviderFactory
from app.services.integration.auth_manager import IntegrationAuthManager
from app.core import logger

logger = logging.getLogger(__name__)    

class IntegrationService:
    """Enhanced service for managing external integrations with validation and improved error handling."""
    
    def __init__(self):
        self.repo = IntegrationRepository()
        self.sync_repo = IntegrationSyncRepository()
        self.provider_factory = ProviderFactory()
        self.auth_manager = IntegrationAuthManager()
        
    def get_available_integrations(self) -> List[Dict[str, Any]]:
        """Get list of available integration providers with updated information."""
        return [
            {
                "id": "shopify",
                "name": "Shopify",
                "description": "E-commerce platform integration for products, orders, and customers",
                "resource_types": ["products", "orders", "customers", "inventory_levels", "locations"],
                "auth_types": ["access_token"],
                "fields": [
                    {
                        "name": "store_domain",
                        "label": "Store Domain",
                        "type": "text",
                        "placeholder": "your-store.myshopify.com",
                        "required": True,
                        "help": "Your Shopify store domain (e.g., your-store.myshopify.com)"
                    },
                    {
                        "name": "access_token",
                        "label": "Admin API Access Token",
                        "type": "password",
                        "placeholder": "shpat_...",
                        "required": True,
                        "help": "Admin API Access Token from your Shopify Custom App"
                    }
                ],
                "setup_instructions": [
                    "Go to your Shopify Admin → Settings → Apps and sales channels",
                    "Click 'Develop apps' → 'Create an app'",
                    "Configure Admin API access with required scopes",
                    "Generate Admin API access token",
                    "Copy the token (starts with 'shpat_') and your store domain"
                ]
            },
            {
                "id": "zendesk",
                "name": "Zendesk",
                "description": "Customer service and engagement platform",
                "resource_types": ["tickets", "users", "organizations"],
                "auth_types": ["api_key", "oauth2"],
                "fields": [
                    {
                        "name": "subdomain",
                        "label": "Zendesk Subdomain",
                        "type": "text",
                        "placeholder": "yourcompany",
                        "required": True,
                        "help": "Your Zendesk subdomain (from yourcompany.zendesk.com)"
                    },
                    {
                        "name": "api_key",
                        "label": "API Token",
                        "type": "password",
                        "required": True,
                        "help": "API token from Zendesk Admin → Channels → API"
                    },
                    {
                        "name": "email",
                        "label": "Admin Email",
                        "type": "email",
                        "required": True,
                        "help": "Email address of Zendesk admin user"
                    }
                ]
            },
            {
                "id": "salesforce",
                "name": "Salesforce",
                "description": "CRM platform integration",
                "resource_types": ["contacts", "accounts", "opportunities", "leads"],
                "auth_types": ["oauth2"],
                "fields": [
                    {
                        "name": "instance_url",
                        "label": "Salesforce Instance",
                        "type": "text",
                        "placeholder": "https://yourinstance.salesforce.com",
                        "required": True,
                        "help": "Your Salesforce instance URL"
                    },
                    {
                        "name": "client_id",
                        "label": "Consumer Key",
                        "type": "text",
                        "required": True,
                        "help": "Consumer Key from Connected App"
                    },
                    {
                        "name": "client_secret",
                        "label": "Consumer Secret",
                        "type": "password",
                        "required": True,
                        "help": "Consumer Secret from Connected App"
                    }
                ]
            }
        ]
    
    def get_client_integrations(
        self, db: Session, client_id: str
    ) -> List[Dict[str, Any]]:
        """Get all integrations for a client with enhanced details."""
        integrations = self.repo.get_by_client_id(db, client_id)
        
        result = []
        for integration in integrations:
            # Get latest sync info
            latest_sync = self.sync_repo.get_latest_by_integration_id(db, integration.integration_id)
            
            # Format integration data
            integration_data = {
                "integration_id": integration.integration_id,
                "provider": integration.provider,
                "name": integration.name,
                "status": integration.status,
                "status_message": integration.status_message,
                "endpoint_url": integration.endpoint_url,
                "is_active": integration.is_active,
                "last_sync": integration.last_sync.isoformat() if integration.last_sync else None,
                "created_at": integration.created_at.isoformat(),
                "updated_at": integration.updated_at.isoformat(),
                "config": {
                    # Only return non-sensitive config data
                    key: value for key, value in (integration.config or {}).items() 
                    if key not in ["credentials", "api_key", "access_token", "client_secret"]
                }
            }
            
            # Add sync information
            if latest_sync:
                integration_data["latest_sync"] = {
                    "status": latest_sync.status,
                    "start_time": latest_sync.start_time.isoformat(),
                    "end_time": latest_sync.end_time.isoformat() if latest_sync.end_time else None,
                    "items_processed": latest_sync.items_processed,
                    "items_created": latest_sync.items_created,
                    "items_updated": latest_sync.items_updated,
                    "items_failed": latest_sync.items_failed
                }
            
            result.append(integration_data)
        
        return result
    
    def test_integration_connection(
        self,
        db: Session,
        client_id: str,
        provider: str,
        credentials: Dict[str, Any],
        config: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """Test integration connection with comprehensive validation."""
        try:
            # Get provider implementation
            provider_impl = self.provider_factory.get_provider(provider)
            
            # Extract endpoint from config or credentials
            endpoint_url = None
            if config:
                endpoint_url = config.get("endpoint_url") or config.get("store_domain") or config.get("subdomain")
            
            if not endpoint_url and "store_domain" in credentials:
                endpoint_url = credentials["store_domain"]
            elif not endpoint_url and "subdomain" in credentials:
                endpoint_url = credentials["subdomain"]
            
            # Test connection
            result = provider_impl.test_connection(
                endpoint_url=endpoint_url,
                credentials=credentials,
                config=config
            )
            
            # Log test result
            logger.info(f"Integration test for {provider} - Client {client_id}: {result['success']}")
            
            return result
            
        except Exception as e:
            logger.exception(f"Error testing integration connection for {provider}: {str(e)}")
            return {
                "success": False,
                "message": f"Connection test failed: {str(e)}",
                "status_code": 500
            }
    
    def create_integration(
        self,
        db: Session,
        client_id: str,
        provider: str,
        name: str,
        credentials: Dict[str, Any],
        config: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """Create a new integration with enhanced validation."""
        try:
            # Check if integration already exists for this client and provider
            existing = self.repo.get_by_provider(db, client_id, provider)
            if existing:
                raise ValueError(f"Integration with provider {provider} already exists for this client")
            
            # Test connection first
            test_result = self.test_integration_connection(db, client_id, provider, credentials, config)
            if not test_result["success"]:
                return {
                    "success": False,
                    "message": f"Connection test failed: {test_result['message']}",
                    "error_code": "CONNECTION_FAILED"
                }
            
            # Prepare configuration
            if config is None:
                config = {}
            
            # Store credentials securely in config
            config["credentials"] = credentials
            
            # Extract endpoint URL from credentials or config
            endpoint_url = None
            if "store_domain" in credentials:
                endpoint_url = credentials["store_domain"]
            elif "subdomain" in credentials:
                endpoint_url = f"{credentials['subdomain']}.zendesk.com"
            elif "instance_url" in credentials:
                endpoint_url = credentials["instance_url"]
            elif config and "endpoint_url" in config:
                endpoint_url = config["endpoint_url"]
            
            # Create integration
            integration_data = {
                "client_id": client_id,
                "provider": provider,
                "name": name,
                "status": "connected",
                "status_message": "Successfully connected",
                "endpoint_url": endpoint_url,
                "config": config,
                "is_active": True
            }
            
            integration = self.repo.create(db, obj_in=integration_data)
            
            # Log successful creation
            logger.info(f"Integration created successfully: {integration.integration_id} for client {client_id}")
            
            # Return success response
            return {
                "success": True,
                "message": "Integration created successfully",
                "integration": {
                    "integration_id": integration.integration_id,
                    "provider": integration.provider,
                    "name": integration.name,
                    "status": integration.status,
                    "endpoint_url": integration.endpoint_url,
                    "is_active": integration.is_active,
                    "created_at": integration.created_at.isoformat()
                }
            }
            
        except ValueError as e:
            logger.warning(f"Validation error creating integration: {str(e)}")
            return {
                "success": False,
                "message": str(e),
                "error_code": "VALIDATION_ERROR"
            }
        except Exception as e:
            logger.exception(f"Error creating integration: {str(e)}")
            return {
                "success": False,
                "message": f"Failed to create integration: {str(e)}",
                "error_code": "CREATION_ERROR"
            }
    
    def update_integration(
        self,
        db: Session,
        integration_id: str,
        update_data: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Update an existing integration with validation."""
        try:
            # Get existing integration
            integration = self.repo.get_by_integration_id(db, integration_id)
            if not integration:
                return {
                    "success": False,
                    "message": f"Integration not found: {integration_id}",
                    "error_code": "NOT_FOUND"
                }
            
            # If credentials are being updated, test connection
            if "credentials" in update_data or "config" in update_data:
                # Merge new credentials with existing ones
                new_credentials = update_data.get("credentials", {})
                existing_credentials = integration.config.get("credentials", {}) if integration.config else {}
                merged_credentials = {**existing_credentials, **new_credentials}
                
                # Test connection with new credentials
                test_result = self.test_integration_connection(
                    db, 
                    integration.client_id, 
                    integration.provider, 
                    merged_credentials,
                    update_data.get("config", integration.config)
                )
                
                if not test_result["success"]:
                    return {
                        "success": False,
                        "message": f"Connection test failed: {test_result['message']}",
                        "error_code": "CONNECTION_FAILED"
                    }
                
                # Update config with new credentials
                new_config = integration.config.copy() if integration.config else {}
                new_config["credentials"] = merged_credentials
                if "config" in update_data:
                    new_config.update(update_data["config"])
                update_data["config"] = new_config
                update_data["status"] = "connected"
                update_data["status_message"] = "Successfully updated and connected"
            
            # Update integration
            updated_integration = self.repo.update(db, db_obj=integration, obj_in=update_data)
            
            logger.info(f"Integration updated successfully: {integration_id}")
            
            return {
                "success": True,
                "message": "Integration updated successfully",
                "integration": {
                    "integration_id": updated_integration.integration_id,
                    "provider": updated_integration.provider,
                    "name": updated_integration.name,
                    "status": updated_integration.status,
                    "endpoint_url": updated_integration.endpoint_url,
                    "is_active": updated_integration.is_active,
                    "updated_at": updated_integration.updated_at.isoformat()
                }
            }
            
        except Exception as e:
            logger.exception(f"Error updating integration: {str(e)}")
            return {
                "success": False,
                "message": f"Failed to update integration: {str(e)}",
                "error_code": "UPDATE_ERROR"
            }
    
    def delete_integration(self, db: Session, integration_id: str) -> Dict[str, Any]:
        """Delete an integration with proper cleanup."""
        try:
            integration = self.repo.get_by_integration_id(db, integration_id)
            if not integration:
                return {
                    "success": False,
                    "message": f"Integration not found: {integration_id}",
                    "error_code": "NOT_FOUND"
                }
            
            # Deactivate first
            self.repo.deactivate(db, integration_id)
            
            # Delete the integration (this will cascade to related syncs)
            self.repo.delete(db, id=integration.id)
            
            logger.info(f"Integration deleted successfully: {integration_id}")
            
            return {
                "success": True,
                "message": "Integration deleted successfully"
            }
            
        except Exception as e:
            logger.exception(f"Error deleting integration: {str(e)}")
            return {
                "success": False,
                "message": f"Failed to delete integration: {str(e)}",
                "error_code": "DELETE_ERROR"
            }
    
    def get_integration_data(
        self,
        db: Session,
        client_id: str,
        provider: str,
        resource_type: str,
        query: Optional[str] = None,
        filters: Optional[Dict[str, Any]] = None
    ) -> List[Dict[str, Any]]:
        """Get data from integration with enhanced error handling."""
        try:
            # Get active integration for this provider
            integration = self.repo.get_by_provider(db, client_id, provider)
            if not integration or not integration.is_active:
                logger.warning(f"No active integration found for provider {provider} and client {client_id}")
                return []
            
            return self.get_data(integration, resource_type, query, filters)
            
        except Exception as e:
            logger.exception(f"Error getting integration data: {str(e)}")
            return []
    
    def get_data(
        self,
        integration: Integration,
        resource_type: str,
        query: Optional[str] = None,
        filters: Optional[Dict[str, Any]] = None
    ) -> List[Dict[str, Any]]:
        """Get data from an integration with comprehensive error handling."""
        try:
            # Get provider implementation
            provider = self.provider_factory.get_provider(integration.provider)
            
            # Extract credentials from config
            credentials = integration.config.get("credentials", {}) if integration.config else {}
            
            if not credentials:
                logger.error(f"No credentials found for integration {integration.integration_id}")
                return []
            
            # Fetch data
            data = provider.get_data(
                integration.endpoint_url,
                credentials,
                resource_type,
                query,
                filters,
                integration.config
            )
            
            # Update last sync time
            self.repo.update_last_sync(db=None, integration_id=integration.integration_id)
            
            return data
            
        except Exception as e:
            logger.exception(f"Error fetching data from integration {integration.integration_id}: {str(e)}")
            # Update integration status to indicate error
            try:
                self.repo.update_status(
                    db=None, 
                    integration_id=integration.integration_id, 
                    status="error", 
                    status_message=f"Data fetch error: {str(e)}"
                )
            except:
                pass  # Don't fail the main operation if status update fails
            return []